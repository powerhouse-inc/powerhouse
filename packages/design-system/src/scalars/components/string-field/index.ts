export * from "./string-field.js";
