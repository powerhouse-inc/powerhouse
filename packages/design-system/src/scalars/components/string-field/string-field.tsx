import React from "react";
import {
  TextField,
  type TextFieldProps,
} from "../fragments/text-field/index.js";
import {
  TextareaField,
  type TextareaFieldProps,
} from "../fragments/textarea-field/index.js";

export interface StringFieldProps
  extends Omit<TextFieldProps, keyof TextareaFieldProps>,
    TextareaFieldProps {}

export const StringField = React.forwardRef<
  HTMLInputElement | HTMLTextAreaElement,
  StringFieldProps
>(({ autoExpand, multiline, ...props }, ref) => {
  if (autoExpand || multiline) {
    // only textarea supports autoExpand and multiline
    return (
      <TextareaField
        autoExpand={autoExpand}
        multiline={multiline}
        {...(props as TextareaFieldProps)}
        ref={ref as React.ForwardedRef<HTMLTextAreaElement>}
      />
    );
  }

  return (
    <TextField
      {...(props as TextFieldProps)}
      ref={ref as React.ForwardedRef<HTMLInputElement>}
    />
  );
});

StringField.displayName = "StringField";
