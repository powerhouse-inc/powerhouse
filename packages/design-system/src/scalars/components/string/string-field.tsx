import { forwardRef } from "react";
import { Input, InputProps } from "../input";
import { FormLabel } from "../form-label";
import { FormMessageList } from "../form-message";
import { FormGroup } from "../form-group";
import { FieldCommonProps } from "../types";

export interface StringProps
  extends FieldCommonProps<string>,
    Omit<InputProps, "value"> {}

export const StringField = forwardRef<HTMLInputElement, StringProps>(
  (
    {
      label,
      description,
      value,
      default: defaultValue,
      errors,
      warnings,
      ...props
    },
    ref,
  ) => {
    return (
      <FormGroup>
        {label && (
          <FormLabel
            required={props.required}
            description={description}
            hasError={!!errors?.length}
          >
            {label}
          </FormLabel>
        )}
        <Input
          value={value ?? undefined}
          defaultValue={defaultValue ?? undefined}
          {...props}
          ref={ref}
        />
        {warnings && <FormMessageList messages={warnings} type="warning" />}
        {errors && <FormMessageList messages={errors} type="error" />}
      </FormGroup>
    );
  },
);
