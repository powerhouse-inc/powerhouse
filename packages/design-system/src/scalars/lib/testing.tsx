import { render, type RenderResult } from "@testing-library/react";
import { Form } from "../components/index.js";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
