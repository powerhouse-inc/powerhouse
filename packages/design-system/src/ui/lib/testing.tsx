import { Form } from "@powerhousedao/design-system";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
