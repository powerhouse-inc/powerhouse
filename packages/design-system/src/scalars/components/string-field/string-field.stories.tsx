import type { Meta, StoryObj } from "@storybook/react";
import { withForm } from "../../lib/decorators.js";
import {
  getDefaultArgTypes,
  getValidationArgTypes,
  PrebuiltArgTypes,
  StorybookControlCategory,
} from "../../lib/storybook-arg-types.js";
import { StringField } from "./string-field.js";

const meta: Meta<typeof StringField> = {
  title: "Document Engineering/Scalars/String Field",
  component: StringField,
  decorators: [withForm, (Story) => <div className="w-48">{Story()}</div>],
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  argTypes: {
    ...getDefaultArgTypes(),
    ...PrebuiltArgTypes.placeholder,
    ...PrebuiltArgTypes.autoComplete,
    ...PrebuiltArgTypes.spellCheck,

    multiline: {
      control: "boolean",
      description: "Whether to render as a multi-line textarea",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    autoExpand: {
      control: "boolean",
      description:
        "Whether textarea should automatically expand as content grows",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    ...PrebuiltArgTypes.trim,
    ...PrebuiltArgTypes.uppercase,
    ...PrebuiltArgTypes.lowercase,
    ...PrebuiltArgTypes.diffMode,
    ...PrebuiltArgTypes.viewMode,
    ...PrebuiltArgTypes.baseValue,

    ...getValidationArgTypes(),
    ...PrebuiltArgTypes.minLength,
    ...PrebuiltArgTypes.maxLength,
    ...PrebuiltArgTypes.pattern,
  },
  args: {
    name: "string-field",
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    label: "String Input",
    placeholder: "Type something...",
  },
};

export const WithDescription: Story = {
  args: {
    label: "Bio",
    description: "Tell us a little bit about yourself",
    placeholder: "I am a...",
  },
};

export const Required: Story = {
  args: {
    label: "Full Name",
    required: true,
    placeholder: "John Doe",
  },
};

export const WithValue: Story = {
  args: {
    label: "Email",
    value: "john.doe@example.com",
    placeholder: "Enter your email",
  },
};

export const Disabled: Story = {
  args: {
    label: "Read Only Field",
    value: "This field is disabled",
    disabled: true,
    placeholder: "This field is disabled",
  },
};

export const WithError: Story = {
  args: {
    label: "Username",
    value: "123",
    required: true,
    errors: ["Username must be at least 8 characters long"],
    placeholder: "Enter your username",
  },
};

export const WithWarning: Story = {
  args: {
    label: "Password",
    value: "password123",
    warnings: ["Consider using a stronger password"],
    placeholder: "Enter your password",
  },
};

export const Multiline: Story = {
  args: {
    label: "Comments",
    multiline: true,
    placeholder: "Enter your comments here...",
    autoExpand: true,
  },
};

export const WithDefaultValue: Story = {
  args: {
    label: "Favorite Color",
    defaultValue: "Blue",
    placeholder: "Enter your favorite color",
  },
};
