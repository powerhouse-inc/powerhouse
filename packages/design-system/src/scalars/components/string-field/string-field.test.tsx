import { screen, waitFor } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";
import { renderWithForm } from "../../lib/testing.js";
import { StringField } from "./string-field.js";

describe("StringField", () => {
  it("should match snapshot with minimal props", () => {
    const { container } = renderWithForm(<StringField name="test" />);
    expect(container).toMatchSnapshot();
  });

  it("should match snapshot when multiline is true (textarea)", () => {
    const { container } = renderWithForm(<StringField name="test" multiline />);
    expect(container).toMatchSnapshot();
  });

  it("should render label when provided", () => {
    renderWithForm(<StringField name="test" label="Test Label" />);
    expect(screen.getByText("Test Label")).toBeInTheDocument();
  });

  it("should render description when provided", () => {
    renderWithForm(<StringField name="test" description="Test description" />);
    expect(screen.getByText("Test description")).toBeInTheDocument();
  });

  it("should render error messages when provided", async () => {
    renderWithForm(<StringField name="test" errors={["Error message"]} />);
    await waitFor(() =>
      expect(screen.getByText("Error message")).toBeInTheDocument(),
    );
  });

  it("should render warning messages when provided", () => {
    renderWithForm(<StringField name="test" warnings={["Warning message"]} />);
    expect(screen.getByText("Warning message")).toBeInTheDocument();
  });

  it("should render as textarea when multiline is true", () => {
    renderWithForm(<StringField name="test" multiline />);
    expect(screen.getByRole("textbox")).toHaveProperty("tagName", "TEXTAREA");
  });

  it("should render as input when multiline is false", () => {
    renderWithForm(<StringField name="test" />);
    expect(screen.getByRole("textbox")).toHaveProperty("tagName", "INPUT");
  });

  it("should disable input when disabled prop is true", () => {
    renderWithForm(<StringField name="test" disabled />);
    expect(screen.getByRole("textbox")).toBeDisabled();
  });

  it("should show required indicator when required prop is true", () => {
    renderWithForm(<StringField name="test" label="Test Label" required />);
    expect(screen.getByText("*")).toBeInTheDocument();
  });

  it("should handle value changes", async () => {
    const user = userEvent.setup();
    const onChange = vi.fn();
    renderWithForm(
      <StringField name="test" value="initial" onChange={onChange} />,
    );
    const input = screen.getByRole("textbox");

    expect(input).toHaveValue("initial");
    await user.type(input, "test");
    expect(onChange).toHaveBeenCalledTimes(4); // Once per character typed
  });
});
