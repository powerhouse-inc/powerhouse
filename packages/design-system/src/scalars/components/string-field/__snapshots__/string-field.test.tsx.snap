// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StringField > should match snapshot when multiline is true (textarea) 1`] = `
<div>
  <form
    novalidate=""
  >
    <div
      class="flex flex-col gap-2"
    >
      <div
        class="relative"
      >
        <textarea
          aria-invalid="false"
          autocomplete="off"
          class="flex w-full rounded-md text-[14px] font-normal leading-5 dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white font-sans placeholder:text-gray-600 dark:placeholder:text-gray-500 px-3 py-[7.2px] focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-1 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 disabled:cursor-not-allowed disabled:border-gray-200 disabled:text-gray-500 resize-y scrollbar-thin scrollbar-track-transparent scrollbar-thumb-gray-300 dark:scrollbar-track-gray-900 dark:scrollbar-thumb-gray-600"
          id=":r1:-textarea"
          name="test"
          rows="3"
        />
      </div>
    </div>
  </form>
</div>
`;

exports[`StringField > should match snapshot with minimal props 1`] = `
<div>
  <form
    novalidate=""
  >
    <div
      class="flex flex-col gap-2"
    >
      <input
        class="flex h-9 w-full rounded-md text-sm text-gray-900 dark:text-gray-700 border border-gray-300 bg-white dark:border-charcoal-700 dark:bg-charcoal-900 px-3 py-2 font-sans placeholder:text-gray-600 dark:placeholder:text-gray-500 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-1 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 disabled:cursor-not-allowed disabled:border-gray-200 disabled:text-gray-500"
        id=":r0:"
        name="test"
        value=""
      />
    </div>
  </form>
</div>
`;
